src/core.php

```diff
@@ -1101,3 +1226,29 @@ public function __construct(:x:composable-element $that, int $index) {
 interface XHPUnsafeRenderable extends XHPChild {
   public function toHTMLString();
 }
+
+/**
+ * INCREDIBLY AWESOME: Specify an element as awaitable on render.
+ *
+ * This allows you to use generators inside your XHP objects. For
+ * instance, you could fetch data inside your XHP elements using await and
+ * the calls to the DB would be batched together when the element is rendered.
+ */
+interface XHPAwaitable {
+  // protected function genRender(): Awaitable<XHPRoot>
+}
+
+trait XHPGenerator {
+
+  require extends :x:element;
+  require implements XHPAwaitable;
+
+  abstract protected function genRender(): Awaitable<XHPRoot>;
+
+  final protected function render(): XHPRoot {
+    throw new Exception(
+      'You need to call genRender() on XHP elements that use XHPAwaitable',
+    );
+    return <x:frag />;
```

Comment:
The problem is that `render()` is declared abstract in `:x:element`, so it needs to be done somewhere. This is why I originally tried to just use `render()` instead of `genRender()` but it munged the return type.